sdk/core/azure-core/inc/http/request.hpp

```diff
@@ -10,138 +10,129 @@
 #include <string>
 #include <vector>
 
-#include <internal/contract.hpp>
-
-namespace azure
-{
-namespace core
-{
-namespace http
-{
-
-class Request
-{
-
-private:
-  Request(Request const&) = delete;
-  void operator=(Request const&) = delete;
-
-  // query needs to be first or at least before url, since url might update it
-  std::map<std::string, std::string> _queryParameters;
-
-  HttpMethod _method;
-  std::string _url;
-  std::map<std::string, std::string> _headers;
-  std::map<std::string, std::string> _retryHeaders;
-  std::map<std::string, std::string> _retryQueryParameters;
-  // Request can contain no body, or either of next bodies (_bodyBuffer plus size or bodyStream)
-  BodyStream* _bodyStream;
-  BodyBuffer* _bodyBuffer;
-
-  // flag to know where to insert header
-  bool _retryModeEnabled;
-
-  // returns left map plus all items in right
-  // when duplicates, left items are preferred
-  static std::map<std::string, std::string> mergeMaps(
-      std::map<std::string, std::string> const& left,
-      std::map<std::string, std::string> const& right)
-  {
-    auto result = std::map<std::string, std::string>(left);
-    std::for_each(right.begin(), right.end(), [&](auto const& pair) { result.insert(pair); });
-    return result;
-  }
-
-  /**
-   * Create an stream from source string. Then use getline and separator to get as many tokens as
-   * possible and insert each one to result vector
-   */
-  static std::vector<std::string> split(std::string source, char separator)
-  {
-    auto result = std::vector<std::string>();
+#include "internal/contract.hpp"
+#include "http/http.hpp"
 
-    auto stringAsStream = std::stringstream(source);
-    auto token = std::string();
+namespace Azure { namespace Core { namespace Http {
 
-    while (std::getline(stringAsStream, token, separator))
+  class Request
+  {
+
+  private:
+    Request(Request const&) = delete;
+    void operator=(Request const&) = delete;
+
+    // query needs to be first or at least before url, since url might update it
+    std::map<std::string, std::string> _queryParameters;
+
+    HttpMethod _method;
+    std::string _url;
+    std::map<std::string, std::string> _headers;
+    std::map<std::string, std::string> _retryHeaders;
+    std::map<std::string, std::string> _retryQueryParameters;
+    // Request can contain no body, or either of next bodies (_bodyBuffer plus size or bodyStream)
+    BodyStream* _bodyStream;
```

Comment:
I understand the request body can be a buffer or a stream, and you want to handle both cases. But I really don't think this is a good idea to do like this. Because we're maintaining two sets of logic all the way from the outmost API into innermost code.

My proposal is storage team wraps the buffer into a stream, and then we only need to deal with stream. We do so in cpplite sdk and have the wrapper off-the-shelf, it works like a charm. See it [here](https://github.com/Azure/azure-storage-cpplite/blob/master/include/mstream.h).

Reply:
ditto for response.